refactor(topology/*): define and use dense_inducing

PatrickMassot opened the pull request:
Traditionally, topology extends functions defined on dense subspaces (equipped by the induced topology). In mathlib, this was made type-theory-friendly by rather factoring through `dense_embedding` maps. A map `f : α  → β` between topological spaces is a dense embedding if its image is dense, it is injective, and it pulls back the topology from `β` to the topology on `α`. It turns out
that the injectivity was never used in any serious way. It used not to be used at all until we noticed it could be used to ensure the factorization equation `dense_embedding.extend_e_eq` could be made to
hold without any assumption on the map to be extended. But of course this formalization trick is mathematically completely irrelevant.

On the other hand, assuming injectivity prevents direct use in uniform spaces completion, because the map from a space to its (separated) completion is injective only when the original space is separated. This is why mathlib ring completion currently assumes a separated topological ring, and the perfectoid spaces project needed a lot of effort to drop that assumption. This commit makes all this completely painless.

Along the way, we improve consistency and readability by turning a couple of conjunctions into structures. Also we introduce a predicate `dense_range` which was already everywhere, but without a name, and with duplicated proofs.

It also introduces a long overdue fix to `function.uncurry` (which suffered from abusive pattern
matching, similar to `prod.map`). Also removes the duplication `nhds_induced_eq_comap`/`nhds_induced`.

@sgouezel could you check I didn't harm your metric space completion and Gromov-Hausdorff work? It should be only little simplification, or a bit of help to the elaborator (especially a couple of `: _` added with speed).

@rwbarton same question for the Stone-Cech compactification. The main difference should be that the map from a discrete space to ultrafilter on it is first proved to be dense inducing before adding that injectivity also holds. And then injectivity is no longer used anywhere. It means the `t2_space` assumption needs to be added a bit earlier but this should be totally irrelevant.

See discussion at https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/function.20extension and https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/induced.20neighborhoods

jcommelin commented on src/data/prod.lean:
```diff
@@ -78,3 +78,8 @@ instance lex.decidable [decidable_eq α] [decidable_eq β]
 end prod
+
+lemma function.injective_prod {f : α → γ} {g : β → δ}
+  (hf : ∀ {a a'}, f a = f a' → a = a') (hg : ∀ {b b'}, g b = g b' → b = b') :
```
Why do you not use `function.injective` in the assumptions and conclusion?

PatrickMassot replied:
I thought this needed more imports (this file is at the very bottom of mathlib, importing nothing) but I was wrong: `function.injective` is in core. I'll fix that.

jcommelin commented on src/topology/algebra/uniform_group.lean:
```diff
@@ -354,8 +342,8 @@ variables [topological_space β] [add_comm_group β] [topological_add_group β]
 variables [uniform_space G] [add_comm_group G] [uniform_add_group G] [separated G] [complete_space G]
-variables {e : β → α} [is_add_group_hom e] (de : dense_embedding e)
-variables {f : δ → γ} [is_add_group_hom f] (df : dense_embedding f)
+variables {e : β → α} [is_add_group_hom e] (de : dense_inducing e)
```
Should these maps now be renamed to `f` and `g`? After all you no longer assume that `e` is an *embedding*. And then `de` and `df` become `df` and `dg` respectively.
I do understand that this kind of refactoring is not so nice with our current tools )-;

PatrickMassot replied:
I don't think it's worth the trouble.

jcommelin commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -13,35 +13,46 @@ local attribute [instance] classical.prop_decidable
+((dense_inducing_coe.prod dense_inducing_coe).extend_eq_of_cont
+  ((continuous_coe α).comp continuous_mul') (a, b)).symm
+
+lemma continuous_mul' : continuous (λp:completion α×completion α, p.1 * p.2) :=
```
```suggestion
lemma continuous_mul' : continuous (λ p : completion α × completion α, p.1 * p.2) :=
```
I find this slightly more readable...

jcommelin commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -78,27 +89,48 @@ instance : ring (completion α) :=
-    [complete_space β] [separated β]
-  {f : α → β} [is_ring_hom f] (hf : continuous f) :
+
+universes u v
```
```suggestion
universes u
```
I don't think you are using `v` anywhere.

jcommelin commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -78,27 +89,48 @@ instance : ring (completion α) :=
     (assume a b,
       by rw [← coe_mul, extension_coe hf, extension_coe hf, extension_coe hf, is_ring_hom.map_mul f]) }
 
+instance top_ring_compl : topological_ring (completion α) :=
```
Nice!

jcommelin commented on src/topology/constructions.lean:
```diff
@@ -338,8 +344,7 @@ lemma continuous_sum_rec {f : α → γ} {g : β → γ}
 lemma embedding_inl : embedding (@sum.inl α β) :=
-⟨λ _ _, sum.inl.inj_iff.mp,
-  begin
+⟨⟨begin
```
I would prefer using named fields instead of anonymous constructor notation...

PatrickMassot replied:
Good catch. I tried to remove those, but missed this pair

jcommelin commented on src/topology/constructions.lean:
```diff
@@ -353,11 +358,10 @@ lemma embedding_inl : embedding (@sum.inl α β) :=
 lemma embedding_inr : embedding (@sum.inr α β) :=
-⟨λ _ _, sum.inr.inj_iff.mp,
-  begin
+⟨⟨begin
```
Idem

jcommelin commented on src/topology/uniform_space/completion.lean:
```diff
@@ -142,20 +142,23 @@ mem_uniformity.trans $ bex_congr $ λ t h, prod.forall
 ⟨pure a, cauchy_pure⟩
 
+lemma uniform_inducing_pure_cauchy : uniform_inducing (pure_cauchy : α → Cauchy α) :=
+⟨have (preimage (λ (x : α × α), (pure_cauchy (x.fst), pure_cauchy (x.snd))) ∘ gen) = id,
```
Here I think you can also fight mathlib's obfuscation by using named fields instead of anonymous constructor notation.

PatrickMassot replied:
I think in single-field structure this is not so bad.

jcommelin commented on src/topology/maps.lean:
```diff
@@ -13,130 +13,200 @@ local attribute [instance] classical.prop_decidable
-function.injective f ∧ tα = tβ.induced f
+structure embedding [tα : topological_space α] [tβ : topological_space β] (f : α → β)
+  extends inducing f : Prop :=
+(inj  :  function.injective f)
```
```suggestion
(inj : function.injective f)
```

jcommelin commented on src/topology/maps.lean:
```diff
@@ -148,111 +218,126 @@ variables [topological_space δ] {f : γ → α} {g : γ → δ} {h : δ → β}
+
+structure dense_embedding [topological_space α] [topological_space β] (e : α → β)
+  extends dense_inducing e : Prop  :=
+(inj   : function.injective e)
```
```suggestion
(inj : function.injective e)
```

jcommelin reviewed (commented):
Some minor nitpicks. Looks really good!

sgouezel commented on src/topology/algebra/uniform_ring.lean:
```diff
@@ -13,35 +13,46 @@ local attribute [instance] classical.prop_decidable
-instance : has_mul (completion α) :=
-⟨λa b, extend (dense_embedding_coe.prod dense_embedding_coe)
-  ((coe : α → completion α) ∘ (λp:α×α, p.1 * p.2)) (a, b)⟩
+instance : has_mul (completion α) := ⟨curry $ (dense_inducing_coe.prod dense_inducing_coe).extend
```
line break after `:=` (and merge with the next line)

sgouezel commented on src/topology/maps.lean:
```diff
@@ -13,130 +13,200 @@ local attribute [instance] classical.prop_decidable
+
+variables {f}
+
+lemma dense_range_comp (hf : dense_range f) (hg : dense_range g) (cg : continuous g) :
```
```suggestion
lemma dense_range.comp (hg : dense_range g) (hf : dense_range f) (cg : continuous g) :
```
the convention for composition is that the order in the assumptions should match the order in `g o f`. Also use a name with dot for easier use.

PatrickMassot replied:
I think the dot name won't do anything, because `dense_range` is not a structure. I could make it a structure, but I was afraid to go overboard with structures for properties of maps. Maybe I'll do it at some point.

sgouezel replied:
It's surprising at first but extremely useful: you don't need structures for the dot notation. If you have an assumption `h : dense_range f`, `h.comp` will be understood as `dense_range.comp h` even if `dense_range` is not a structure.

sgouezel commented on src/topology/maps.lean:
```diff
@@ -13,130 +13,200 @@ local attribute [instance] classical.prop_decidable
+  exact hg c
+end
+
+lemma dense_range.inhabited (df : dense_range f) (b : β) : inhabited α :=
```
Why `inhabited` instead of `nonempty`? since there is no canonical element, I think `nonempty` is more natural -- and if you really need `inhabited` for some applications, maybe it means that the statements you rely on should also use `nonempty` instead of `inhabited`, or otherwise you can easily derive `inhabited` from `nonempty` in the application.

PatrickMassot replied:
`inhabited` is needed because this is what `lim` wants.

sgouezel commented on src/topology/maps.lean:
```diff
@@ -13,130 +13,200 @@ local attribute [instance] classical.prop_decidable
 
 variables [topological_space α] [topological_space β] [topological_space γ] [topological_space δ]
 
+def embedding.mk' (f : α → β) (inj  :  function.injective f)
```
```suggestion
def embedding.mk' (f : α → β) (inj : function.injective f)
```

sgouezel commented on src/topology/maps.lean:
```diff
@@ -13,130 +13,200 @@ local attribute [instance] classical.prop_decidable
 lemma embedding.closure_eq_preimage_closure_image {e : α → β} (he : embedding e) (s : set α) :
   closure s = e ⁻¹' closure (e '' s) :=
-by ext x; rw [set.mem_preimage, ← closure_induced he.1, he.2]
+by { ext x,  rw [set.mem_preimage, ← closure_induced he.inj, he.induced] }
```
```suggestion
by { ext x, rw [set.mem_preimage, ← closure_induced he.inj, he.induced] }
```

sgouezel commented on src/topology/maps.lean:
```diff
@@ -13,130 +13,200 @@ local attribute [instance] classical.prop_decidable
--- TODO: use embeddings from above!
-structure dense_embedding [topological_space α] [topological_space β] (e : α → β) : Prop :=
+structure dense_inducing [topological_space α] [topological_space β] (e : α → β)
+  extends inducing e : Prop  :=
```
```suggestion
  extends inducing e : Prop :=
```

sgouezel commented on src/topology/maps.lean:
```diff
@@ -13,130 +13,200 @@ local attribute [instance] classical.prop_decidable
+namespace dense_inducing
 variables [topological_space α] [topological_space β]
-variables {e : α → β} (de : dense_embedding e)
+variables {e : α → β} (di : dense_inducing e)
```
I don't like this name `di` since it refers to `e`. Or you could rename `e` to `i` everywhere: this is probably more natural, but maybe too painful for what it's worth.

PatrickMassot replied:
Right, I'll switch to `i`. It's not fun to do, but we don't want historical accident to mess up notations.

sgouezel commented on src/topology/maps.lean:
```diff
@@ -148,111 +218,126 @@ variables [topological_space δ] {f : γ → α} {g : γ → δ} {h : δ → β}
+  di.extend f b = c :=
+@lim_eq _ _ (id _) _ _ _ (by simp; exact comap_nhds_neq_bot di) hf
+
+lemma extend_e_eq [t2_space γ] {f : α → γ} (a : α)  (hf : continuous_at f a) :
```
```suggestion
lemma extend_e_eq [t2_space γ] {f : α → γ} (a : α) (hf : continuous_at f a) :
```

sgouezel commented on src/topology/metric_space/completion.lean:
```diff
@@ -54,25 +49,25 @@ protected lemma completion.dist_comm (x y : completion α) : dist x y = dist y x
 begin
   apply induction_on₃ x y z,
-  { refine is_closed_le _ (continuous_add _ _),
+   { refine is_closed_le _ (continuous_add _ _),
```
what happened here?
```suggestion
  { refine is_closed_le _ (continuous_add _ _),
```

sgouezel commented on src/topology/uniform_space/basic.lean:
```diff
@@ -766,6 +758,26 @@ lemma to_topological_space_prod [u : uniform_space α] [v : uniform_space β] :
+
+lemma uniform_continuous₂.comp {f : α → β → γ} {g : γ → δ}
+  (hg : uniform_continuous g) (hf : uniform_continuous₂ f) :
+uniform_continuous₂ (g ∘₂ f) :=
```
```suggestion
  uniform_continuous₂ (g ∘₂ f) :=
```

sgouezel commented on src/topology/uniform_space/uniform_embedding.lean:
```diff
@@ -1,29 +1,56 @@
-def uniform_embedding [uniform_space β] (f : α → β) :=
-function.injective f ∧
-comap (λx:α×α, (f x.1, f x.2)) (𝓤 β) = 𝓤 α
+lemma uniform_inducing.comp {f : α → β} (hf : uniform_inducing f)
```
Can you change the order of assumptions on `g` and `f`?

sgouezel commented on src/topology/uniform_space/uniform_embedding.lean:
```diff
@@ -1,29 +1,56 @@
+structure uniform_embedding (f : α → β) extends uniform_inducing f : Prop :=
+(inj : function.injective f)
+
+lemma uniform_embedding.comp {f : α → β} (hf : uniform_embedding f)
```
Can you change the order of assumptions on g and f?

sgouezel commented on src/topology/uniform_space/uniform_embedding.lean:
```diff
@@ -1,29 +1,56 @@
     ∃ t ∈ 𝓤 β, ∀ x y : α, (f x, f y) ∈ t → (x, y) ∈ s :=
-by rw [uniform_embedding, eq_comm, filter.ext_iff]; simp [subset_def]
+begin
+  split, rintro ⟨⟨h⟩, h'⟩,
```
I did not open the proof in Lean, but a proof starting with split and without braces looks suspicious.

PatrickMassot committed:
refactor(topology/*): define and dense_inducing

Traditionally, topology extends functions defined on dense subspaces
(equipped by the induced topology).
In mathlib, this was made type-theory-friendly by rather factoring
through `dense_embedding` maps. A map `f : α  → β` between topological
spaces is a dense embedding if its image is dense, it is injective, and
it pulls back the topology from `β` to the topology on `α`. It turns out
that the injectivity was never used in any serious way. It used not to
be used at all until we noticed it could be used to ensure the
factorization equation `dense_embedding.extend_e_eq` could be made to
hold without any assumption on the map to be extended. But of course
this formalization trick is mathematically completely irrelevant.

On the other hand, assuming injectivity prevents direct use in uniform
spaces completion, because the map from a space to its (separated)
completion is injective only when the original space is separated. This
is why mathlib ring completion currently assumes a separated topological
ring, and the perfectoid spaces project needed a lot of effort to drop
that assumption. This commit makes all this completely painless.

Along the way, we improve consistency and readability by turning
a couple of conjunctions into structures. It also introduces long
overdue fix to `function.uncurry` (which suffered from abusive pattern
matching, similar to `prod.map`).

PatrickMassot reviewed (commented):
Thank you very much @sgouezel and @jcommelin!

PatrickMassot committed:
Apply suggestions from code review

Co-Authored-By: Johan Commelin <johan@commelin.net>
Co-Authored-By: sgouezel <sebastien.gouezel@univ-rennes1.fr>

PatrickMassot reviewed (commented):
I think I've fixed everything, and this is good to go.

sgouezel commented on src/topology/maps.lean:
```diff
@@ -13,133 +13,203 @@ local attribute [instance] classical.prop_decidable
+lemma inducing_id : inducing (@id α) :=
+⟨induced_id.symm⟩
+
+lemma inducing_compose {f : α → β} {g : β → γ} (hg : inducing g) (hf : inducing f) :
```
```suggestion
lemma inducing.comp {f : α → β} {g : β → γ} (hg : inducing g) (hf : inducing f) :
```

jcommelin replied:
There are more `compose` lemmas in this area of mathlib. Maybe we should rename them all in one go in a separate PR.

PatrickMassot replied:
I fixed three of them.

sgouezel commented on src/topology/maps.lean:
```diff
@@ -13,133 +13,203 @@ local attribute [instance] classical.prop_decidable
+  inducing (g ∘ f) :=
+⟨by rw [hf.induced, hg.induced, induced_compose]⟩
+
+lemma inducing_prod_mk {f : α → β} {g : γ → δ} (hf : inducing f) (hg : inducing g) :
```
```suggestion
lemma inducing.prod_mk {f : α → β} {g : γ → δ} (hf : inducing f) (hg : inducing g) :
```

sgouezel commented on src/topology/maps.lean:
```diff
@@ -13,133 +13,203 @@ local attribute [instance] classical.prop_decidable
 end embedding
 
+structure dense_inducing [topological_space α] [topological_space β] (i : α → β)
+  extends inducing i : Prop  :=
```
```suggestion
  extends inducing i : Prop :=
```

sgouezel commented on src/topology/maps.lean:
```diff
@@ -13,133 +13,203 @@ local attribute [instance] classical.prop_decidable
+by rw [continuous_at, di.nhds_eq_comap a]; exact tendsto_comap
 
-protected lemma continuous (de : dense_embedding e) : continuous e :=
+protected lemma continuous (de : dense_inducing i) : continuous i :=
```
You forgot to rename `de` here.

PatrickMassot committed:
minor fixes following review

mergify[bot] committed:
Merge branch 'master' into extension

PatrickMassot committed:
Some more dot notation, consistent naming and field naming
